lib/dry/system/settings.rb

```diff
@@ -38,19 +39,30 @@ def self.setting(*args)
 
         def self.load(root, env)
           env_data = load_files(root, env)
+          errors = {}
 
           attributes = schema.each_with_object({}) do |(key, type), h|
             value = ENV.fetch(key.to_s.upcase) { env_data[key.to_s.upcase] }
+            check_for_error(key, type, value, errors)
```

Comment:
The way `check_for_error` works here makes me feel a bit uncomfortable: it's inside a loop mutating `h` to provide a meaningful return value, while along the way you're passing a _different_ object to to be mutated by some other method far away. Just doesn't clean to me.

I could see us cleaning it up in a couple of ways:

1. Initializing `attributes = {}` above and turning this into an `each` loop, then assigning to `attributes` and `errors` (if required) for each setting. This is just being honest to ourselves that this particular code is procedural :)
2. Assigning `attributes` as we do currently (with the `check_for_error` line removed), then adding _another_ `each_with_index` afterwards to build up an errors hash. This would be a bit slower though because we're looping through the attributes twice.

(1) is probably fine here, I think. What do you reckon?

Reply:
Tim totally agree with you, it doesn't look clean. 

Between both suggestions, I would go for the first option, fewer iterations.